### PR TITLE
Bir remake

### PR DESCRIPTION
**Pull request by joaobrittoneto**

### CODE REVIEW

File: src/Driver.cpp [hunk 1]
```diff
@@ -1,598 +1,530 @@
...
-        std::cout << "buffer_as_string: " << buffer_as_string << std::endl;
-        if (UsblParser::isPacket(buffer_as_string) > 0){
-            if (handleAsynchronousCommand(buffer_as_string)){
-                packet_size = 0;
+int Driver::extractPacket(uint8_t const* buffer, size_t buffer_size) const
+{
+	std::string buffer_as_string = std::string(reinterpret_cast<char const*>(buffer), buffer_size);
+	buffer_as_string = buffer_as_string.substr(0, buffer_size);
+
+	int ret;
```
Comment (doudou) [inline src/Driver.cpp]: Don't do that. Define your local variables the closest to where they are needed.

**Comment by doudou** [comment]
One comment: you use a mix of tabs and spaces, and have tabsize != 4. This fucks up indentation on github (and on everyone that does not have the same tab size than you ... that is everyone using the default for most editors, 8 spaces per tab). Please only use spaces to indent (and reindent this ...)

File: src/UsblParser.hpp [hunk 2]
```diff
@@ -3,114 +3,43 @@
...
-             * @param[in] size The length of data
-             * @param[in] command The command to fit the data with this command 
-             * @return The extracted string
-             */
-            static std::string parseString(uint8_t const* data, size_t const size, std::string const command);
+            ~UsblParser();
+
+//            int isPacket(std::string const &buffer, OperationMode const &mode);
+
+            bool findNotification(std::string const &buffer, Notification &notification);
```
Comment (doudou) [inline src/UsblParser.hpp]: All these booleans are very suspicious-looking. Remove them, throw exceptions if there are errors, return the data that you want to return if there is one (here, obviously Notification).

File: src/UsblParser.hpp [hunk 3]
```diff
@@ -3,114 +3,43 @@
...
+
+
+            bool findResponse(std::string const &buffer, CommandResponse &response);
+            bool validateResponse(std::string const &buffer, std::string const &command);
+            bool validateParticularResponse(std::string const &buffer);
+
+
+            std::string parseSendIM(SendIM const &im);
+
+            std::string parseReceivedIM(std::string const &buffer, ReceiveIM &im);
```
Comment (doudou) [inline src/UsblParser.hpp]: You fill in a structure *and* return a string ?

I think you will definitely have to document all that stuff ...

**Comment by doudou** [comment]
Another one: don't pass reference arguments as a way to return a value, unless you REALLY have no other choice.

**Comment by joaobrittoneto** [comment]
@doudou,
I updated the code according your remarks. 
I avoid pass variables through reference, use exception, improve indentation.

File: src/Driver.cpp [hunk 4]
```diff
@@ -1,598 +1,558 @@
...
-    std::cout << "now returning Y: " << current_position.y << std::endl;
-    std::cout << "now returning Z: " << current_position.z << std::endl;
+WaitResponse Driver::sendCommand(void)
+{
+    if(!queueCommand.empty())
+    {
+        std::string command = queueCommand.front();
 
-        new_position_available = false;
+        std::cout << "Write Line: " << command << std::endl;
```
Comment (doudou) [inline src/Driver.cpp]: Don't check in code that has debug output like this. If you really want to have debug output, use Rock's logger (http://rock-robotics.org/master/documentation/debugging/index.html)

File: src/Driver.cpp [hunk 5]
```diff
@@ -1,598 +1,558 @@
...
+        break;
+    case BUSY:
+        ss << "TRY AGAIN. ";
+        ss << buffer << std::flush;
+        ret = ss.str();
+        break;
+    case COMMAND_RECEIVED:
+        modeMsgManager(command);
+        ss << "OK" << buffer << std::flush;
+        ret = ss.str();
```
Comment (doudou) [inline src/Driver.cpp]: Just "return ss.str();"

File: src/Driver.cpp [hunk 6]
```diff
@@ -1,598 +1,558 @@
...
-            return true;
-        case USBLLONG:
-            incomingPosition(buffer_as_string);
-        case USBLANGLE:
-            std::cout << "There is a USBL Asynch Message" << std::endl;
-            //TODO Handle this, when possible (unknown protocol)
-            return true;
-    }
-    return false;
+void Driver::goSurface(void)
```
Comment (doudou) [inline src/Driver.cpp]: I am pretty sure the USBL evologics does not have a dedicated goSurface command. Leave this out of the library.

File: src/UsblParser.hpp [hunk 7]
```diff
@@ -3,114 +3,72 @@
...
+    std::vector<std::string> splitValidate(std::string const &buffer, const char* symbol, size_t const parts);
+    //            static std::vector<std::string> validateResponse(std::string const s);
+    //            static int getInt(std::string s);
+
+public:
+    UsblParser();
+    ~UsblParser();
+
+
+    // Given buffer, determine if it's a Notification and returns it's kind.
```
Comment (doudou) [inline src/UsblParser.hpp]: These comments are not properly formatted for doxygen. Use

~~~
/** Summary line
 *
 * long comment
 */
~~~

File: src/Driver.cpp [hunk 8]
```diff
@@ -1,598 +1,558 @@
...
+    return NO_RESPONSE_REQUIRED;
 }
 
-int Driver::getSystemTime(){
-    sendWithLineEnding("+++AT?UT");
-    return waitSynchronousInt();
+bool Driver::sendRawData(void)
+{
+    if(!queueRawData.empty())
+    {   // TODO Need to check the size of the Raw data?
```
Comment (doudou) [inline src/Driver.cpp]: If this could be the case, you need to move this comment at the place the data is queued, not where it is consumed.

**Comment by doudou** [comment]
More general comment:
 - what is the rationale behind the command queue ?
 - I can't figure out where the response type is being used ...

File: src/Driver.cpp (hunk 6 above)
Comment (joaobrittoneto) [inline src/Driver.cpp]: Ok. I was thinking about the possibility of sending high level commands to vehicle, through a Instant Message, the same way I'd like to send the pose to vehicle. Please see IMParser.hpp. 
So should I avoid to implement these commands and just do for the pose?

Comment (doudou) [inline src/Driver.cpp]: This is the evologics driver, stick to what the device does natively.

After that, you could create a auv_uwmodem_commands library or something that would sit on top to provide a generic command/status library

**Comment by joaobrittoneto** [comment]
- I use a command queue in order to have a stream of commands to be sent to device. The usbl's manual suggest to wait for a response of a command before send another command to device. So with the queue I could control the flow of command sent.  

- The answer type would be used in the component, in order to know what the device is sending to driver (a response, a notification or raw data) and output it in the proper port. I draft version of it could be viewed in Main.cpp

**Comment by doudou** [comment]
> I use a command queue in order to have a stream of commands to be sent to device. The usbl's manual suggest to wait for a response of a command before send another command to device. So with the queue I could control the flow of command sent.

You would have the exact same effect by waiting for the response in the method that sent the command. Only the code would be a lot simpler. Is there anything that goes against that solution ? 

> The answer type 

I asked for "response type". Some methods return whether responses are expected or not, but I could not see where that would be used.

**Comment by joaobrittoneto** [comment]
- I tried to make independently the send command from the read response, once the device can send back 3 kind of data: a response from a command, a unexpected notification and raw data. So I would read in only one place, then classify the kind of data and interpret it. Notification or raw data can be received between a command and it response and they can be received without any command be sent.  
I realize the code in is getting complex by slit away the send command and read response, but I am not sure how to deal with these unexpected data.

- The device doesn't send a response to command "AT0" (Switch to Data mode). It's the only command that doesn't receive a response. So the get response loop wouldn't wait for a response from this command.

**Comment by doudou** [comment]
> I tried to make independently the send command from the read response, once the device can send back 3 kind of data: a response from a command, a unexpected notification and raw data. So I would read in only one place, then classify the kind of data and interpret it. Notification or raw data can be received between a command and it response and they can be received without any command be sent.

The usual way is to make the response reading a loop that waits for the response, and queues any other message (raw data/notifications) into separate queues. It becomes the responsibility of the caller to empty these queues later on. At least the command/response flow is clear from the API user p.o.v.

**Comment by joaobrittoneto** [comment]
Suggestions implemented.

**Pull request merged**